snuba_sdk/schema.py

```diff
@@ -0,0 +1,38 @@
+from dataclasses import dataclass
+from typing import Sequence, Set
+
+
+@dataclass(frozen=True)
+class Column:
+    name: str
+
+
+@dataclass(frozen=True, init=False)
+class ColumnSet:
+    columns: Set[str]
+
+    def __init__(self, columns: Sequence[Column]) -> None:
+        super().__setattr__("columns", set(c.name for c in columns))
+
+    def contains(self, column_name: str) -> bool:
+        return column_name in self.columns
+
+
+@dataclass(frozen=True, init=False)
+class EntityModel:
+    column_set: ColumnSet
+    required_time_column: Column
+    required_columns: Sequence[Column]
```

Comment:
I find this API a bit unintuitive.

- How come `required_time_column` is passed separately to required columns? Isn't just a type of required column? Also, how come it is mandatory? Although our entities currently do all have a time column, I don't think it is a requirement we want to impose on every entity. In Snuba itself, it is actually optional. Also is the required time column also supposed to get passed into `required_columns` and `column_set` ? It's not very clear.

- I think it's a bit weird to pass in the same column multiple times. Couldn't we just have a property on the column indicating whether it's required or not?

Reply:
The time column is separate because we have to validate it differently. It requires two conditions instead of one, and with specific ops as well. It should be optional, since I think CDC entities don't require it.

You do make an interesting point about making "required" a property of the column instead of a separate property of the data model.